Add a function isIPAddressInRange()

depressed-pho opened the pull request:
This function takes an IP address and a CIDR network prefix, and determines if the address is contained in the network. It supports both IPv4 and IPv6 addresses, and expects arguments to be strings.

ClickHouse doesn't support union types. We cannot use the native IPv4 and IPv6 types for this reason.

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Add function `isIPAddressInRange` to test if an IPv4 or IPv6 address is contained in a given CIDR network prefix.

Detailed description / Documentation draft:
Contained in the commit.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

depressed-pho committed:
Add a function isIPAddressContainedIn()

This function takes an IP address and a CIDR network prefix, and determines if the address is contained in the network. It supports both IPv4 and IPv6 addresses, and expects arguments to be strings.

ClickHouse doesn't support union types. We cannot use the native IPv4 and IPv6 types for this reason.

alexey-milovidov commented:
Please take a look at the reports of fuzzers.

alexey-milovidov commented:
```
WITH
    CAST('127.0.0.1', 'Nullable(String)') AS addr,
    arrayJoin([NULL, NULL, NULL, NULL]) AS prefix
SELECT
    '1\07.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/81\07.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8127.0.0.0/8',
    isIPAddressContainedIn('127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1127.0.0.1', addr, prefix)
```

depressed-pho commented:
> Please take a look at the reports of fuzzers.

Yes, it was an actual bug in my code. It should be fixed now.

vdimir commented on src/Functions/isIPAddressContainedIn.cpp:
```diff
@@ -0,0 +1,739 @@
+            }
+            else
+            {
+                if (bits == 128)
```
We have function to convert integer prefix length to binary representation
https://github.com/ClickHouse/ClickHouse/blob/e27715e55e471dc4d7f825eb4097b468a9a4a451/src/Common/IPv6ToBinary.cpp#L67

It statically generate all of 128 possible masks. But maybe this implementation with `if (bits >= ...) ... else if ...` event better, not sure. Anyway, we can leave one implementation.

depressed-pho replied:
Okay, I will reuse that function. Thanks for the feedback.

vdimir commented:
We already have some functions that overlaps with some implemented in this PR.

<details>
<summary>See IPAddressDictionary.cpp and FunctionsCoding.h
</summary>


https://github.com/ClickHouse/ClickHouse/blob/e27715e55e471dc4d7f825eb4097b468a9a4a451/src/Dictionaries/IPAddressDictionary.cpp#L83-L243

https://github.com/ClickHouse/ClickHouse/blob/e27715e55e471dc4d7f825eb4097b468a9a4a451/src/Functions/FunctionsCoding.h#L1646-L1669

</details>

Can you consider reusing this functions?

We can move common functionality to work with IP addresses to common library like https://github.com/ClickHouse/ClickHouse/blob/e27715e55e471dc4d7f825eb4097b468a9a4a451/src/Common/IPv6ToBinary.cpp



> ClickHouse doesn't support union types. We cannot use the native IPv4 and IPv6 types for this reason.

Why we cant have overload for `String`/`IPv4`/`IPv6` types? I think now it's ok to support one string type, but later we can support columns with other types too.

depressed-pho committed:
Fix build on GCC where std::isinf() cannot be called without the namespace prefix

depressed-pho committed:
Correct comments about what parseIPv4() and parseIPv6() actually do

They were wrong as to what endian their results were in.

depressed-pho commented:
> We already have some functions that overlaps with some implemented in this PR.
> See IPAddressDictionary.cpp and FunctionsCoding.h
> 
> Could you consider reusing this functions?

Yes. Will do.

> > ClickHouse doesn't support union types. We cannot use the native IPv4 and IPv6 types for this reason.
> 
> Why we cant have overload for `String`/`IPv4`/`IPv6` types? I think now it's ok to support one string type, but later we can support columns with other types too.

In the future we can do that of course, but for now I'm going to only support strings.

depressed-pho committed:
Changed the return type of DB::getCIDRMaskIPv6()

It now returns std::array instead of uint8_t *. Also added DB::getCIDRMaskIPv4().

qoega commented on docs/en/sql-reference/functions/ip-address-functions.md:
````diff
@@ -394,3 +394,41 @@ Result:
+Query:
+
+``` sql
+SELECT isIPAddressContainedIn('127.0.0.1', '127.0.0.0/8')
````
Probably we need an example or separate note what happens if we try to find IPv4 in IPv6 subnet?

depressed-pho replied:
Okay, will write one.

vdimir committed:
Reuse some functions for IPAddressContainedIn

vdimir commented:
@depressed-pho 

Thank you for your PR.
I read code, all seems good.

I rewrite part of functions dealing with IPv4/IPv6 to reuse some code from IPAddressDictionary it's slightly decreased size of code (see commit 24aa25d7dcb982079713394078c621c2c671695c).
Feel free to comment my changes.

Also I think we can do not support nullable columns, because function have no special behavior for `NULLs`, so it's up to user to handle nullable columns. I'm going to remove code handling `Nullable`. What do you think?

vdimir committed:
Remove Nummable support from isIPAddressContainedIn, do not use OverloadResolver

depressed-pho commented:
> I rewrite part of functions dealing with IPv4/IPv6 to reuse some code from IPAddressDictionary it's slightly decreased size of code (see commit [24aa25d](https://github.com/ClickHouse/ClickHouse/commit/24aa25d7dcb982079713394078c621c2c671695c)).
> Feel free to comment my changes.

Thanks. All of your changes look good to me.

> Also I think we can do not support nullable columns, because function have no special behavior for `NULLs`, so it's up to user to handle nullable columns. I'm going to remove code handling `Nullable`. What do you think?

I don't understand well. I thought functions in ClickHouse were usually supposed to return NULL if any of their arguments were NULL, so I did the same for `isIPAddressContainedIn()`. If you think it's unnecessary or inappropriate, I will agree on the removal.

vdimir committed:
Rename isIPAddressContainedIn -> isIPAddressInRange

vdimir committed:
Add test cases for isIPAddressInRange

vdimir committed:
Rename test file isIPAddressContainedIn to ip_address_in_range

vdimir commented:
> I don't understand well. I thought functions in ClickHouse were usually supposed to return NULL if any of their arguments were NULL, so I did the same for `isIPAddressContainedIn()`

I meant that user can handle Nullable with `coalesce` of `ifNull`. Without supporting Nullable code is much less over-complicated. Let's remove Nullable support for now. We always can return it back if it will be necessary.

vdimir committed:
Upd doc for isIPAddressInRange

vdimir committed:
Fix build isIPAddressContainedIn

gyuton commented:
Internal documentation ticket: DOCSUP-8181.

filimonov commented:
@depressed-pho  do you plan to add support for IPv4 / IPv6 datatypes?

alexey-milovidov commented:
A deficiency has been found: #34796
